### PR TITLE
feat(python, rust): clearer error message when replace_time_zone encounters ambiguous or non-existent datetimes

### PR DESCRIPTION
**Pull request by MarcoGorelli**
This came up on discord

On master (shows neither the offending datetime, nor the offending time zone):
```python
In [91]: pl.Series(['2021-03-28 02:00']).str.strptime(pl.Datetime).dt.replace_time_zone('Europe/Warsaw')
thread '<unnamed>' panicked at 'No such local time', /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/chrono-0.4.23/src/offset/mod.rs:186:34
---------------------------------------------------------------------------

PanicException: No such local time
```

Here:
```python
In [1]: pl.Series(['2021-03-28 02:00']).str.strptime(pl.Datetime).dt.replace_time_zone('Europe/Warsaw')
---------------------------------------------------------------------------

ArrowErrorException: InvalidArgumentError("datetime '2021-03-28 02:00:00' is non-existent in time zone 'Europe/Warsaw'. Non-existent datetimes are not yet supported")
```

note: I've removed irrelevant parts of the traceback here for readability

### CODE REVIEW

**Comment by ritchie46** [comment]
Nice! I have a small remark that may help remove some boilerplate code/types.

**Comment by MarcoGorelli** [comment]
that's much better, thanks! learning every day
Just trying to figure out how to use it in `polars-arrow` without having a circular import

**Comment by ritchie46** [comment]
> Just trying to figure out how to use it in polars-arrow without having a circular import

The earliest (lowest in the dag) crate should implement it. So in this case I would put the trait definition in the polars-arrow crate and export it.

**Comment by MarcoGorelli** [comment]
Sure, thanks - this require marking `chrono` as non-optional in `polars-arrow` though, right? Because there are places where the type needs to be annotated as `Optional<&impl PolarsTimeZone>` which aren't behind `feature = "timezones"`, e.g.

https://github.com/pola-rs/polars/blob/2d17b8e61ce67977bedb5bf91407ddc9c4d37b70/polars/polars-time/src/windows/duration.rs#L327-L336

File: polars/polars-arrow/Cargo.toml [hunk 1]
```diff
@@ -10,7 +10,7 @@ description = "Arrow interfaces for Polars DataFrame library"
 
 [dependencies]
 arrow.workspace = true
-chrono = { version = "0.4", default-features = false, features = ["std"], optional = true }
+chrono = { version = "0.4", default-features = false, features = ["std"] }
```
Comment (ritchie46) [inline polars/polars-arrow/Cargo.toml:13]: Could we make this optional. In that case `PolarsTimeZone` will be:

```rust
#[not(cfg(feature="timezones"))]
pub trait PolarsTimeZone: std::fmt::Debug + std::fmt::Display {}

#[cfg(feature="timezones")]
pub trait PolarsTimeZone: chrono::TimeZone + std::fmt::Debug + std::fmt::Display {}
```

Comment (MarcoGorelli) [reply polars/polars-arrow/Cargo.toml:13]: thanks! I've tried:
```Rust
#[cfg(feature = "timezones")]
use chrono_tz::Tz;

#[cfg(not(feature = "timezones"))]
pub trait PolarsTimeZone: std::fmt::Debug + std::fmt::Display {}

#[cfg(feature = "timezones")]
pub trait PolarsTimeZone: chrono::TimeZone + std::fmt::Debug + std::fmt::Display {}

#[cfg(feature = "timezones")]
impl PolarsTimeZone for chrono::FixedOffset {}
#[cfg(feature = "timezones")]
impl PolarsTimeZone for Tz {}
```
and this passes the tests, but then `make pre-commit` fails with
```

```Rust
error[E0277]: the trait bound `chrono::FixedOffset: polars_arrow::PolarsTimeZone` is not satisfied
  --> polars/polars-time/src/date_range.rs:60:60
   |
60 |             date_range_vec(start, stop, every, closed, tu, NO_TIMEZONE)?,
   |             --------------                                 ^^^^^^^^^^^ the trait `polars_arrow::PolarsTimeZone` is not implemented for `chrono::FixedOffset`
   |             |
   |             required by a bound introduced by this call
```

Comment (MarcoGorelli) [reply polars/polars-arrow/Cargo.toml]: for now I've reverted to defining `PolarsTimeZone` in `polars-time/src/lib.rs` and just adding `Debug` and `Display` in two places in `polars-arrow`

if I figure out how to define it in `polars-arrow` I'll add another commit

Comment (ritchie46) [reply polars/polars-arrow/Cargo.toml]: Could you make it work until you are stuck? Then I can try to make a PR on top of this PR. I think it should work.

File: polars/polars-arrow/src/time_zone.rs [hunk 2]
```diff
@@ -0,0 +1,23 @@
+use std::fmt::{Debug, Display};
+#[cfg(not(feature = "timezones"))]
+pub trait PolarsTimeZone: Debug + Display + Sync + Send + Clone {}
+
+// a dummy type that implements required subtraits
+#[cfg(not(feature = "timezones"))]
+impl PolarsTimeZone for u8 {}
+#[cfg(not(feature = "timezones"))]
+pub const NO_TIMEZONE: Option<&u8> = None;
```
Comment (ritchie46) [inline polars/polars-arrow/src/time_zone.rs:9]: Ensure we have a `NO_TIMEZONE` implemented in both cases.

File: polars/polars-time/Cargo.toml [hunk 3]
```diff
@@ -31,7 +31,7 @@ dtype-duration = ["polars-core/dtype-duration", "polars-core/temporal"]
 rolling_window = ["polars-core/rolling_window", "dtype-duration"]
 private = []
 fmt = ["polars-core/fmt"]
-timezones = ["chrono-tz", "dtype-datetime", "polars-core/timezones"]
+timezones = ["chrono-tz", "dtype-datetime", "polars-core/timezones", "polars-arrow/timezones"]
```
Comment (ritchie46) [inline polars/polars-time/Cargo.toml:34]: Make sure we activate the proper version if we compile with timezones. This was the root of the problem @MarcoGorelli

Comment (MarcoGorelli) [reply polars/polars-time/Cargo.toml:34]: aah nice, thanks Ritchie!

**Pull request merged**